tests/HTMLPurifier/HTMLModule/HTML5/LinkTest.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+class HTMLPurifier_HTMLModule_HTML5_LinkTest extends BaseTestCase
+{
+    protected function setUp()
+    {
+        parent::setUp();
+
+        $this->config->set('Attr.AllowedRel', array('stylesheet', 'preload'));
+        $this->config->set('URI.SafeLinkRegexp', '%^https?://localhost/%');
+    }
+
+    public function testClosingTag()
+    {
+        $this->assertPurification(
+            '<link></link>',
+            ''
+        );
+    }
+
+    public function testSelfClosing()
+    {
+        $this->assertPurification(
+            '<link>',
+            ''
+        );
+    }
+
+    public function testNoHref()
+    {
+        $this->assertPurification(
+            '<link rel="stylesheet">',
+            ''
+        );
+    }
+
+    public function testNoRel()
+    {
+        $this->assertPurification(
+            '<link href="foo">',
+            ''
+        );
+    }
+
+    public function testInvalidUrl()
+    {
+        $this->assertPurification(
+            '<link href="http://google.com/foo.css" rel="stylesheet">',
+            ''
+        );
+    }
+
+    public function testGood()
+    {
+        $this->assertPurification(
+            '<link href="https://localhost/foo.css" rel="stylesheet">',
+            '<link href="https://localhost/foo.css" rel="stylesheet">'
+        );
+    }
+
+    public function testWhitelistCapitalised()
+    {
+        // regex is case sensitive
+        $this->assertPurification(
+            '<link href="https://localHost/FOO.css" rel="stylesheet">',
+            ''
+        );
+    }
+
+    public function testGoodWithAutoclosedTag()
+    {
+        $this->assertPurification(
+            '<link href="https://localhost/foo.css" rel="stylesheet"></link>',
+            '<link href="https://localhost/foo.css" rel="stylesheet">'
+        );
+
+        $this->assertPurification(
+            '<link href="https://localhost/foo.css" rel="stylesheet" />',
+            '<link href="https://localhost/foo.css" rel="stylesheet">'
+        );
+    }
+
+    public function testMultiRel()
+    {
+        $this->assertPurification(
+            '<link href="https://localhost/foo.css" rel="stylesheet preload">',
+            '<link href="https://localhost/foo.css" rel="stylesheet preload">'
+        );
```

Comment:
[NuValidator](https://validator.w3.org/nu/) reports error on this:

```
Error: A link element with a rel attribute that contains the value preload must have an as attribute.

From line 7, column 1; to line 7, column 64
d>↩<body>↩<link href="https://localhost/foo.css" rel="stylesheet preload">↩</bod
```

> Error: A link element with a rel attribute that contains the value preload must have an as attribute.

Until the spec behavior is implemented `preload` value (and possibly some other rel values) shouldn't be supported.

Reply:
Closed as duplicate comment. Will change this to only allow `rel="stylesheet"` per https://github.com/xemlock/htmlpurifier-html5/pull/63/files#r634730866